server/etcdserver/v3_server.go

```diff
@@ -800,6 +802,15 @@ func (s *EtcdServer) requestCurrentIndex(leaderChangedNotifier <-chan struct{},
 			readIndexFailed.Inc()
 			// return a retryable error.
 			return 0, ErrLeaderChanged
```

Comment:
Off-topic (for potential experimenting in the future):  
   I wonder whether with all the support for 'retrying', whether this line is providing any 'good' to the e2e experience. 
I would assume that now the system will self-recover. Although it will require testing.

Reply:
I'd say that there are 3 PRs (#12780, #12762 and current) that are solving issue described in https://github.com/etcd-io/etcd/issues/12680 (lag on read-only requests after leader change). Each of them would solve the issue alone, but all of them are introducing something that others are not. Scenario I can imagine that this one is solving and others are not is following:

- Leader change happens (no committed entry in new term)
- Follower gets read request from client
- Follower sends ReadIndex request to Leader
- Network issue happens and ReadIndex request is not delivered to Leader
- Leader commit first entry in this term

Thanks to this change follower will resend ReadIndex request immediately instead waiting 500ms (#12780) or 7s (#12762). I agree that we talk about deep edge cases, however i feel that it still provides value.

Reply:
I'm sorry. I was not clear. I was not questioning whether we need all 3 changes. 

I was kind of suggesting the 4th one: 
   Currently the outstanding calls when leadership changes are refused and user gets 'ErrLeaderChanged' error. 
   The looking-forward suggestion I'm making:  "Why do we need to return that error at all instead of waiting for successful retry by the first leader ?"

Reply:
Oh, I see. I was not paying attention to code you've mentioned. Indeed that's good question. The only drawback of this proposal I can spot is that read request during leader change will be slower as it will need to get committed index before responding. Ofc faster response means error and potential need to retry on client side, however feedback loop seems tighter in this case.